### ISSUE TITLE
Cyclic dependencies in let's do not throw compiler errors and compile incorrectly

### ISSUE THREAD

**Issue by rhaps0dy** [elm-lang/elm-compiler#1021]
The following code:

```elm
module Main ( main ) where

import Graphics.Element as Element

value : String
value =
  let
    a = "Hello, " ++ b
    b = "World!" ++ a
  in
    a ++ b

main : Signal Element.Element
main = Signal.constant <| Element.show value
```

Compiles, and its output is `"Hello, undefinedWorld!Hello, undefined"`

It should throw an error, something like "cyclic dependencies in let have no clear compilation", or at least a warning.

**Comment by jvoigtlaender** [comment]
Sure this is not "just" another instance of [this issue](https://github.com/elm-lang/elm-compiler/issues/873)?

**Comment by rhaps0dy** [comment]
You are right. It is not the same issue, but it fits in the category of issues that are referenced at the bottom.

**Comment by evancz** [comment]
This is definitely a duplicate of #873, follow along there to keep up with progress on this issue. Thanks for the report!
